### PR TITLE
introduce query type selection

### LINKED ISSUE #403: Option to run / exclude single or all query type
### Please describe your feature request:

dnsx has a dedicated option for each query type which works well for queries specific queries, but not convenient when looking to query all the supported queries or running all and excluding few of them. 

These new options work as an alternate way to select/exclude a type of DNS query to run, not a replacement for existing options.

```console
   -q, -query-type      dns query type to check (a,aaaa,cname,ns,txt,srv,ptr,mx,soa,axfr,caa,any,all)
   -eq, -exclude-type   dns query type to exclude (a,aaaa,cname,ns,txt,srv,ptr,mx,soa,axfr,caa,any,all)
```

### Describe the use case of this feature:

Here is an example of running dnsx to query all supported dns queries against domain or list of domains and exclude few of them as needed.

```console
echo google.com | dnsx -q all -eq ptr,axfr,any
dnsx -list dns.txt -q all -eq ptr,axfr,any
```

Note:
>1.  List of supported query types should be populated automatically to ensure any new type supported in the future will be reflected accordingly instead of adding them each time.
> 2. Default / JSONL output will be the same as before.

### PR DESCRIPTION
**Pull request by dogancanbakir**
### Proposal:

dnsx has a dedicated option for each query type which works well for queries specific queries, but not convenient when looking to query all the supported queries or running all and excluding few of them. 

These new options work as an alternate way to select/exclude a type of DNS query to run, not a replacement for existing options.

```console
   -q, -query-type      dns query type to check (a,aaaa,cname,ns,txt,srv,ptr,mx,soa,axfr,caa,any,all)
   -eq, -exclude-type   dns query type to exclude (a,aaaa,cname,ns,txt,srv,ptr,mx,soa,axfr,caa,any,all)
```

### use case:

Here is an example of running dnsx to query all supported dns queries against domain or list of domains and exclude few of them as needed.

```console
echo google.com | dnsx -q all -eq ptr,axfr,any
dnsx -list dns.txt -q all -eq ptr,axfr,any
```

Note:
>1.  List of supported query types should be populated automatically to ensure any new type supported in the future will be reflected accordingly instead of adding them each time.
> 2. Default / JSONL output will be the same as before.

Closes  #403.

### CODE REVIEW

File: internal/runner/options.go [hunk 1]
```diff
@@ -339,3 +362,44 @@ func (options *Options) configureResume() error {
...
+			}
+		} else {
+			if val, ok := queryMap[qt]; ok {
+				*val = true
+			}
+		}
+	}
+
+	for _, et := range options.ExcludeType {
+		if et == "all" {
```
Comment (Mzack9999) [inline internal/runner/options.go:395]: I think `all` should not be a recognized value for exclude query types, as it makes the tool not usable

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
- Using both `-q a -eq a` makes the tool anyway sending out an `a` query (standard behavior)
```console
$ echo google.com | go run . -q a -eq a -json -a | jq

      _             __  __
   __| | _ __   ___ \ \/ /
  / _' || '_ \ / __| \  / 
 | (_| || | | |\__ \ /  \ 
  \__,_||_| |_||___//_/\_\

		projectdiscovery.io

[INF] Current dnsx version 1.1.5 (latest)
{
  "host": "google.com",
  "ttl": 180,
  "resolver": [
    "1.0.0.1:53"
  ],
  "a": [
    "142.250.180.174"
  ],
  "all": [
    "google.com.\t180\tIN\tA\t142.250.180.174",
    "\n;; OPT PSEUDOSECTION:\n; EDNS: version 0; flags:; udp: 1232"
  ],
  "status_code": "NOERROR",
  "timestamp": "2023-10-10T09:44:05.497750082+02:00"
}
```
- Imo the `q` option is redundant with single query types flag (`-q a,aaaa => -a -aaaa`), I think it would have been more functional to just add `-all` flag, that was the only missing case. Differently from `httpx` with an ample amount of http protocol possible flags, dns queries are pretty much written in stone.

File: internal/runner/options.go (hunk 1 above)
Comment (dogancanbakir) [reply internal/runner/options.go]: > Using both -q a -eq a makes the tool anyway sending out an a query (standard behavior)
I haven't touched this behavior, should we?

Comment (Mzack9999) [reply internal/runner/options.go]: I think we should keep the current behavior as base, but if `eq` exclude all other possibilities before adding `a` we can inform the user (ref: https://github.com/nmap/nmap/blob/ff92f5bae941058fae4782167fc7cb49b29c9440/scan_lists.cc#L460)

**Review by Mzack9999 — APPROVED** [review]
lgtm!

Note: I'm not sure if it's worth enabling by default json output with multiple queries as the same domain is repeated multiple times:
```console
$ echo google.com | go run . -all

      _             __  __
   __| | _ __   ___ \ \/ /
  / _' || '_ \ / __| \  / 
 | (_| || | | |\__ \ /  \ 
  \__,_||_| |_||___//_/\_\

                projectdiscovery.io

[INF] Current dnsx version 1.1.5 (latest)
google.com
google.com
google.com
google.com
google.com
google.com
google.com
google.com
```

**Review by ehsandeep — CHANGES_REQUESTED** [review]
@dogancanbakir when `-all` option is used, we can default to below format.

```console
google.com [A] [172.217.167.46]
google.com [A] [142.250.206.142]
google.com [AAAA] [2404:6800:4002:82e::200e]
google.com [AAAA] [2404:6800:4002:82c::200e]
google.com [MX] [smtp.google.com]
google.com [NS] [ns1.google.com]
google.com [NS] [ns3.google.com]
google.com [NS] [ns2.google.com]
google.com [NS] [ns4.google.com]
google.com [TXT] [v=spf1 include:_spf.google.com ~all]
google.com [TXT] [ms=e4a68b9ab2bb9670bce15412f62916164c0b20bb]
google.com [TXT] [google-site-verification=wd8n7i1jtntkezj49swvww48f8_9xverev4ob-0hf5o]
google.com [TXT] [atlassian-domain-verification=5yjtmwmji92ewqkx2oxmbad60td9zwon9r6eakvhx6b77zzkfqto8pq9qsknbf4i]
google.com [TXT] [docusign=1b0a6754-49b1-4db5-8540-d2c12664b289]
google.com [TXT] [globalsign-smime-dv=cdyx+xfhuw2wml6/gb8+59bsh31kzur6c1l2bpvqkx8=]
google.com [TXT] [apple-domain-verification=30afibcvsudv2plx]
google.com [TXT] [facebook-domain-verification=22rm551cu4k0ab0bxsw536tlds4h95]
google.com [TXT] [onetrust-domain-verification=de01ed21f2fa4d8781cbc3ffb89cf4ef]
google.com [TXT] [webexdomainverification.8yx6g=6e6922db-e3e6-4a36-904e-a805c28087fa]
```

**Comment by dogancanbakir** [comment]
```console
$ echo google.com | go run . -all -resp

      _             __  __
   __| | _ __   ___ \ \/ /
  / _' || '_ \ / __| \  / 
 | (_| || | | |\__ \ /  \ 
  \__,_||_| |_||___//_/\_\

                projectdiscovery.io

[INF] Current dnsx version 1.1.6 (latest)
google.com [A] [142.250.186.78] 
google.com [A] [142.250.186.174] 
google.com [AAAA] [2a00:1450:4017:804::200e] 
google.com [AAAA] [2a00:1450:4001:82b::200e] 
google.com [MX] [smtp.google.com] 
google.com [NS] [ns4.google.com] 
google.com [NS] [ns2.google.com] 
google.com [NS] [ns3.google.com] 
google.com [NS] [ns1.google.com] 
google.com [SOA] [ns1.google.com] 
google.com [SOA] [dns-admin.google.com] 
google.com [SOA] [ns1.google.com] 
google.com [SOA] [dns-admin.google.com] 
google.com [SOA] [ns1.google.com] 
google.com [SOA] [dns-admin.google.com] 
google.com [SOA] [ns1.google.com] 
google.com [SOA] [dns-admin.google.com] 
google.com [ANY] [142.250.186.78] 
google.com [ANY] [142.250.186.174] 
google.com [ANY] [2a00:1450:4017:804::200e] 
google.com [ANY] [2a00:1450:4001:82b::200e] 
google.com [ANY] [smtp.google.com] 
google.com [ANY] [ns1.google.com] 
google.com [ANY] [dns-admin.google.com] 
google.com [ANY] [ns4.google.com] 
google.com [ANY] [ns2.google.com] 
google.com [ANY] [ns3.google.com] 
google.com [ANY] [google-site-verification=wd8n7i1jtntkezj49swvww48f8_9xverev4ob-0hf5o] 
google.com [ANY] [ms=e4a68b9ab2bb9670bce15412f62916164c0b20bb] 
google.com [ANY] [atlassian-domain-verification=5yjtmwmji92ewqkx2oxmbad60td9zwon9r6eakvhx6b77zzkfqto8pq9qsknbf4i] 
google.com [ANY] [webexdomainverification.8yx6g=6e6922db-e3e6-4a36-904e-a805c28087fa] 
google.com [ANY] [globalsign-smime-dv=cdyx+xfhuw2wml6/gb8+59bsh31kzur6c1l2bpvqkx8=] 
google.com [ANY] [apple-domain-verification=30afibcvsudv2plx] 
google.com [ANY] [facebook-domain-verification=22rm551cu4k0ab0bxsw536tlds4h95] 
google.com [ANY] [docusign=1b0a6754-49b1-4db5-8540-d2c12664b289] 
google.com [ANY] [docusign=05958488-4752-4ef2-95eb-aa7ba8a3bd0e] 
google.com [ANY] [google-site-verification=tv9-dbe4r80x4v0m4u_bd_j9cpojm0nikft0jagjmsq] 
google.com [ANY] [onetrust-domain-verification=de01ed21f2fa4d8781cbc3ffb89cf4ef] 
google.com [ANY] [v=spf1 include:_spf.google.com ~all] 
google.com [ANY] [pki.goog] 
google.com [TXT] [google-site-verification=wd8n7i1jtntkezj49swvww48f8_9xverev4ob-0hf5o] 
google.com [TXT] [ms=e4a68b9ab2bb9670bce15412f62916164c0b20bb] 
google.com [TXT] [atlassian-domain-verification=5yjtmwmji92ewqkx2oxmbad60td9zwon9r6eakvhx6b77zzkfqto8pq9qsknbf4i] 
google.com [TXT] [webexdomainverification.8yx6g=6e6922db-e3e6-4a36-904e-a805c28087fa] 
google.com [TXT] [globalsign-smime-dv=cdyx+xfhuw2wml6/gb8+59bsh31kzur6c1l2bpvqkx8=] 
google.com [TXT] [apple-domain-verification=30afibcvsudv2plx] 
google.com [TXT] [facebook-domain-verification=22rm551cu4k0ab0bxsw536tlds4h95] 
google.com [TXT] [docusign=1b0a6754-49b1-4db5-8540-d2c12664b289] 
google.com [TXT] [docusign=05958488-4752-4ef2-95eb-aa7ba8a3bd0e] 
google.com [TXT] [google-site-verification=tv9-dbe4r80x4v0m4u_bd_j9cpojm0nikft0jagjmsq] 
google.com [TXT] [onetrust-domain-verification=de01ed21f2fa4d8781cbc3ffb89cf4ef] 
google.com [TXT] [v=spf1 include:_spf.google.com ~all] 
google.com [CAA] [pki.goog] 
```

image

**Review by ehsandeep — CHANGES_REQUESTED** [review]
- bug with PTR query for IPs (`echo 1.1.1.1 | ./dnsx -ptr -resp`)
- bug with ANY query in CLI output (`echo hackerone.com | ./dnsx -all`)

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**